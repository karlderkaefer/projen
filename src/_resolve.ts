import { ResolveOptions, IResolvable } from "./file";

function isResolvable(obj: any): obj is IResolvable {
  return (obj as IResolvable).toJSON !== undefined;
}

export function resolve(value: any, options: ResolveOptions = {}): any {
  const args = options.args ?? [];
  const omitEmpty = options.omitEmpty ?? false;

  if (value == null) {
    return value;
  }

  if (isResolvable(value)) {
    const resolved = value.toJSON();
    return resolve(resolved, options);
  }

  // if value is a function, call it and resolve the result.
  if (typeof value === "function") {
    const resolved = value.apply(undefined, args);
    return resolve(resolved, options);
  }

  if (typeof value !== "object") {
    return value;
  }

  if (Array.isArray(value)) {
    if (omitEmpty && value.length === 0) {
      return undefined;
    }
    return value.map((x) => resolve(x, options)).filter((x) => x != null); // filter undefined/null/omitted
  }

  // only allow data types (i.e. objects without constructors)
  if (value.constructor && value.constructor?.name !== "Object") {
    throw new Error(
      `only data types can be resolved. trying to resolve object of type ${value.constructor?.name}`
    );
  }

  const result: any = {};

  for (const [k, v] of Object.entries(value)) {
    const resolved = resolve(v, options);

    // skip undefined values
    if (resolved === undefined) {
      continue;
    }

    result[k] = resolved;
  }

  if (omitEmpty && Object.keys(result).length === 0) {
    return undefined;
  }

  return result;
}
