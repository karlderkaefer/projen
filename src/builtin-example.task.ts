console.log("hello, I am a builtin task");
console.log("second line");
