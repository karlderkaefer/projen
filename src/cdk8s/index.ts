export * from "./auto-discover";
export * from "./cdk8s-app-ts";
export * from "./cdk8s-construct";
export * from "./integration-test";
