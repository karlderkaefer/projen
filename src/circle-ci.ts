import { Component } from './component';
import { Project } from './project';
import { YamlFile } from './yaml';

export interface CircleCiProps {
  readonly orbs?: Record<string, IOrb>;
  readonly description?: string;
  readonly jobs?: string[];
  readonly version: string;
  readonly workflows?: Record<string, IWorkflow>[];
}

export interface IOrb {
  readonly name: string;
  readonly version: string;
  toString(): string;
}

export class Orb implements IOrb {
  readonly name: string;
  readonly version: string;

  constructor(name: string, version: string) {
    this.name = name;
    this.version = version;
  }
  public toString(): string {
    return `${this.name}@${this.version}`;
  }
}

export interface IJob {
  readonly requires: string[];
  readonly context: string[];
  readonly filterBranchOnly: string[];
}

export interface IWorkflow {
  readonly jobs: Record<string, IJob[]>[];
}

function renderCircleCiFile(props?: CircleCiProps): object {
  const orbs : Record<string, string> = {};
  for (const [id, orb] of Object.entries(props?.orbs ?? {})) {
    orbs[id] = `${orb.toString()}`;
  }
  return {
    version: props?.version || '2.1',
    description: props?.description,
    orbs: orbs,
    jobs: props?.jobs,
    workflows: props?.workflows,
  };
}

export class CircleCi extends Component {
  constructor(project: Project, props?: CircleCiProps) {
    super(project);
    new YamlFile(project, 'circle.yml', {
      committed: true,
      readonly: true,
      obj: () => this._synthesizeCircleCi(props),
    });
  }
  _synthesizeCircleCi(props?: CircleCiProps): object {
    return renderCircleCiFile(props);
  }
}

