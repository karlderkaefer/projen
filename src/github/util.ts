export function secretToString(secretName: string): string {
  return `\${{ secrets.${secretName} }}`;
}
