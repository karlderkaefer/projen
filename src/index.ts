export * from './awscdk-app-ts';
export * from './awscdk-construct';
export * from './cdk8s-app-ts';
export * from './cdk8s-construct';
export * from './cdktf-construct';
export * from './component';
export * from './construct-lib';
export * from './dev-env';
export * from './docker-compose';
export * from './eslint';
export * from './file';
export * from './gitpod';
export * from './ignore-file';
export * from './ini';
export * from './jest';
export * from './jsii-project';
export * from './json';
export * from './logger';
export * from './license';
export * from './makefile';
export * from './node-package';
export * from './node-project';
export * from './object-file';
export * from './option-hints';
export * from './project';
export * from './readme';
export * from './sample-file';
export * from './semver';
export * from './textfile';
export * from './toml';
export * from './typescript';
export * from './typescript-config';
export * from './version';
export * from './yaml';
export * from './xmlfile';
export * from './source-code';
export * from './upgrade-dependencies';
export * from './circle-ci';

// export submobules
export * as tasks from './tasks';
export * as web from './web';
export * as github from './github';
export * as vscode from './vscode';
export * as deps from './deps';
export * as json from './json/index'; // <-- we also have json.ts
export * as java from './java';
export * as python from './python';
export * as typescript from './typescript/index'; // <-- we also have typescript.ts
export * as javascript from './javascript';
export * as release from './release';
export * as git from './git';
