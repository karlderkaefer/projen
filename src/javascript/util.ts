import { basename, dirname, extname, join, sep } from "path";

export function renderBundleName(entrypoint: string) {
  const parts = join(entrypoint).split(sep);
  if (parts[0] === "src") {
    parts.shift(); // just remove 'src' if its the first element for ergonomics
  }

  const p = parts.join(sep);
  const dir = dirname(p);
  const base = basename(p, extname(p));
  return join(dir, base);
}
