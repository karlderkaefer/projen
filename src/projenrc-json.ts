import { existsSync, writeFileSync } from "fs";
import { resolve } from "path";
import { Component } from "./component";
import { Project } from "./project";

export interface ProjenrcOptions {
  /**
   * The name of the projenrc file.
   * @default ".projenrc.json"
   */
  readonly filename?: string;
}

/**
 * Sets up a project to use JSON for projenrc.
 */
export class Projenrc extends Component {
  private readonly rcfile: string;

  constructor(project: Project, options: ProjenrcOptions = {}) {
    super(project);

    this.rcfile = options.filename ?? ".projenrc.json";

    // this is the task projen executes when running `projen`
    project.defaultTask?.env("FILENAME", this.rcfile);
    project.defaultTask?.builtin("run-projenrc-json");

    this.generateProjenrc();
  }

  private generateProjenrc() {
    const rcfile = resolve(this.project.outdir, this.rcfile);
    if (existsSync(rcfile)) {
      return; // already exists
    }

    const bootstrap = this.project.initProject;
    if (!bootstrap) {
      return;
    }

    const json = {
      type: bootstrap.fqn,
      ...bootstrap.args,
    };

    writeFileSync(rcfile, JSON.stringify(json, null, 2));
    this.project.logger.info(
      `Project definition file was created at ${rcfile}`
    );
  }
}
