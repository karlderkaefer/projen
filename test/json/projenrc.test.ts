import { renderProjenInitOptions } from "../../src/javascript/render-options";
import { Projenrc } from "../../src/projenrc-json";
import { synthSnapshot, TestProject } from "../util";

test("projenrc.json default project", () => {
  // GIVEN
  const project = new TestProject();

  // WHEN
  new Projenrc(project);

  // THEN
  expect(synthSnapshot(project)).toMatchSnapshot();
});

test("projenrc.json with typed options", () => {
  // GIVEN
  const project = new TestProject(
    renderProjenInitOptions("projen.typescript.TypeScriptProject", {
      staleOptions: {
        issues: {
          daysBeforeStale: 100, // number, nested option
        },
        pullRequest: {
          enabled: false, // boolean, nested option
        },
      },
      name: "@example/foo", // string
      npmAccess: "public", // enum provided as string
    })
  );

  // WHEN
  new Projenrc(project);

  // THEN
  expect(synthSnapshot(project)).toMatchSnapshot();
});
