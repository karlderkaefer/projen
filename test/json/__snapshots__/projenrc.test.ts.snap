// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`projenrc.json default project 1`] = `
Object {
  ".gitattributes": "# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

/.gitattributes linguist-generated
/.github/workflows/pull-request-lint.yml linguist-generated
/.github/workflows/stale.yml linguist-generated
/.gitignore linguist-generated
/.projen/** linguist-generated
/.projen/deps.json linguist-generated
/.projen/files.json linguist-generated
/.projen/tasks.json linguist-generated",
  ".github/workflows/pull-request-lint.yml": "# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

name: pull-request-lint
on:
  pull_request_target:
    types:
      - labeled
      - opened
      - synchronize
      - reopened
      - ready_for_review
      - edited
jobs:
  validate:
    name: Validate PR title
    runs-on: ubuntu-latest
    permissions:
      pull-requests: write
    steps:
      - uses: amannn/action-semantic-pull-request@v3.4.6
        env:
          GITHUB_TOKEN: \${{ secrets.GITHUB_TOKEN }}
        with:
          types: |-
            feat
            fix
            chore
          requireScope: false
",
  ".github/workflows/stale.yml": "# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

name: stale
on:
  schedule:
    - cron: 0 1 * * *
  workflow_dispatch: {}
jobs:
  stale:
    runs-on: ubuntu-latest
    permissions:
      issues: write
      pull-requests: write
    steps:
      - uses: actions/stale@v4
        with:
          days-before-stale: -1
          days-before-close: -1
          days-before-pr-stale: 14
          days-before-pr-close: 2
          stale-pr-message: This pull request is now marked as stale because it hasn't seen activity for a while. Add a comment or it will be closed soon. If you wish to exclude this issue from being marked as stale, add the \\"backlog\\" label.
          close-pr-message: Closing this pull request as it hasn't seen activity for a while. Please add a comment @mentioning a maintainer to reopen. If you wish to exclude this issue from being marked as stale, add the \\"backlog\\" label.
          stale-pr-label: stale
          exempt-pr-labels: backlog
          days-before-issue-stale: 60
          days-before-issue-close: 7
          stale-issue-message: This issue is now marked as stale because it hasn't seen activity for a while. Add a comment or it will be closed soon. If you wish to exclude this issue from being marked as stale, add the \\"backlog\\" label.
          close-issue-message: Closing this issue as it hasn't seen activity for a while. Please add a comment @mentioning a maintainer to reopen. If you wish to exclude this issue from being marked as stale, add the \\"backlog\\" label.
          stale-issue-label: stale
          exempt-issue-labels: backlog
",
  ".gitignore": "# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".
node_modules/
!/.gitattributes
!/.projen/tasks.json
!/.projen/deps.json
!/.projen/files.json
!/.github/workflows/pull-request-lint.yml
!/.github/workflows/stale.yml
",
  ".projen/files.json": Object {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
    "files": Array [
      ".gitattributes",
      ".github/workflows/pull-request-lint.yml",
      ".github/workflows/stale.yml",
      ".gitignore",
      ".projen/deps.json",
      ".projen/files.json",
      ".projen/tasks.json",
    ],
  },
  ".projen/tasks.json": Object {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
    "tasks": Object {
      "build": Object {
        "description": "Full release build",
        "name": "build",
        "steps": Array [
          Object {
            "spawn": "default",
          },
          Object {
            "spawn": "pre-compile",
          },
          Object {
            "spawn": "compile",
          },
          Object {
            "spawn": "post-compile",
          },
          Object {
            "spawn": "test",
          },
          Object {
            "spawn": "package",
          },
        ],
      },
      "compile": Object {
        "description": "Only compile",
        "name": "compile",
      },
      "default": Object {
        "description": "Synthesize project files",
        "env": Object {
          "FILENAME": ".projenrc.json",
        },
        "name": "default",
        "steps": Array [
          Object {
            "builtin": "run-projenrc-json",
          },
        ],
      },
      "eject": Object {
        "description": "Remove projen from the project",
        "env": Object {
          "PROJEN_EJECTING": "true",
        },
        "name": "eject",
        "steps": Array [
          Object {
            "spawn": "default",
          },
        ],
      },
      "package": Object {
        "description": "Creates the distribution package",
        "name": "package",
      },
      "post-compile": Object {
        "description": "Runs after successful compilation",
        "name": "post-compile",
      },
      "pre-compile": Object {
        "description": "Prepare the project for compilation",
        "name": "pre-compile",
      },
      "test": Object {
        "description": "Run tests",
        "name": "test",
      },
    },
  },
  "README.md": "# replace this",
}
`;

exports[`projenrc.json with typed options 1`] = `
Object {
  ".gitattributes": "# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

/.gitattributes linguist-generated
/.github/workflows/pull-request-lint.yml linguist-generated
/.github/workflows/stale.yml linguist-generated
/.gitignore linguist-generated
/.projen/** linguist-generated
/.projen/deps.json linguist-generated
/.projen/files.json linguist-generated
/.projen/tasks.json linguist-generated",
  ".github/workflows/pull-request-lint.yml": "# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

name: pull-request-lint
on:
  pull_request_target:
    types:
      - labeled
      - opened
      - synchronize
      - reopened
      - ready_for_review
      - edited
jobs:
  validate:
    name: Validate PR title
    runs-on: ubuntu-latest
    permissions:
      pull-requests: write
    steps:
      - uses: amannn/action-semantic-pull-request@v3.4.6
        env:
          GITHUB_TOKEN: \${{ secrets.GITHUB_TOKEN }}
        with:
          types: |-
            feat
            fix
            chore
          requireScope: false
",
  ".github/workflows/stale.yml": "# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

name: stale
on:
  schedule:
    - cron: 0 1 * * *
  workflow_dispatch: {}
jobs:
  stale:
    runs-on: ubuntu-latest
    permissions:
      issues: write
      pull-requests: write
    steps:
      - uses: actions/stale@v4
        with:
          days-before-stale: -1
          days-before-close: -1
          days-before-pr-stale: -1
          days-before-pr-close: -1
          days-before-issue-stale: 100
          days-before-issue-close: 7
          stale-issue-message: This issue is now marked as stale because it hasn't seen activity for a while. Add a comment or it will be closed soon. If you wish to exclude this issue from being marked as stale, add the \\"backlog\\" label.
          close-issue-message: Closing this issue as it hasn't seen activity for a while. Please add a comment @mentioning a maintainer to reopen. If you wish to exclude this issue from being marked as stale, add the \\"backlog\\" label.
          stale-issue-label: stale
          exempt-issue-labels: backlog
",
  ".gitignore": "# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".
node_modules/
!/.gitattributes
!/.projen/tasks.json
!/.projen/deps.json
!/.projen/files.json
!/.github/workflows/pull-request-lint.yml
!/.github/workflows/stale.yml
",
  ".projen/files.json": Object {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
    "files": Array [
      ".gitattributes",
      ".github/workflows/pull-request-lint.yml",
      ".github/workflows/stale.yml",
      ".gitignore",
      ".projen/deps.json",
      ".projen/files.json",
      ".projen/tasks.json",
    ],
  },
  ".projen/tasks.json": Object {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
    "tasks": Object {
      "build": Object {
        "description": "Full release build",
        "name": "build",
        "steps": Array [
          Object {
            "spawn": "default",
          },
          Object {
            "spawn": "pre-compile",
          },
          Object {
            "spawn": "compile",
          },
          Object {
            "spawn": "post-compile",
          },
          Object {
            "spawn": "test",
          },
          Object {
            "spawn": "package",
          },
        ],
      },
      "compile": Object {
        "description": "Only compile",
        "name": "compile",
      },
      "default": Object {
        "description": "Synthesize project files",
        "env": Object {
          "FILENAME": ".projenrc.json",
        },
        "name": "default",
        "steps": Array [
          Object {
            "builtin": "run-projenrc-json",
          },
        ],
      },
      "eject": Object {
        "description": "Remove projen from the project",
        "env": Object {
          "PROJEN_EJECTING": "true",
        },
        "name": "eject",
        "steps": Array [
          Object {
            "spawn": "default",
          },
        ],
      },
      "package": Object {
        "description": "Creates the distribution package",
        "name": "package",
      },
      "post-compile": Object {
        "description": "Runs after successful compilation",
        "name": "post-compile",
      },
      "pre-compile": Object {
        "description": "Prepare the project for compilation",
        "name": "pre-compile",
      },
      "test": Object {
        "description": "Run tests",
        "name": "test",
      },
    },
  },
  ".projenrc.json": Object {
    "name": "@example/foo",
    "npmAccess": "public",
    "staleOptions": Object {
      "issues": Object {
        "daysBeforeStale": 100,
      },
      "pullRequest": Object {
        "enabled": false,
      },
    },
    "type": "projen.typescript.TypeScriptProject",
  },
  "README.md": "# replace this",
}
`;
